NetKVM/Common/ParaNdis-TX.h

```diff
@@ -72,6 +79,13 @@ class CNB : public CNdisAllocatableViaHelper<CNB>
     CNB(const CNB&) = delete;
     CNB& operator= (const CNB&) = delete;
 
+#if MAX_FRAGMENTS_IN_ONE_NB
```

Comment:
@ybendito ,Indeed, your method of recommendation is fantastic, I will transform it into the method you mentioned.

Reply:
@ybendito I would like to discuss some implementation details with you.
1. CNdisSharedMemory* m_UsedPages[MAX_PACKET_PAGES]; // or use the list? just for discussion
    ok, I will change it to a list. At the same time, I will remove the CNdisList<CNdisSharedMemory, CLockedAccess, CCountingObject> SendCopyPagesInUsed;. The original reason I considered using an array was that I was concerned about the possibility of an element being inserted into two lists simultaneously, which could cause errors. Now that SendCopyPagesInUsed has been removed, this issue no longer exists.

2.Regarding the struct CExtendedNBStorage, I understand that this structure is only meant to hold members and will not contain related functions（Except for the destructor and constructor）. For example, MapCopyDataToVirtioSGL will not be included in this struct, correct?

3. SCATTER_GATHER_LIST m_ListHeader // SCATTER_GATHER_LIST
SCATTER_GATHER_ELEMENT m_Elements[MAX_PACKET_PAGES]
In your example, I see that there are these two members. Since SCATTER_GATHER_LIST includes a flexible array, it seems, at first glance, that the definition of m_ListHeader should be placed at the end; otherwise, it fail to compile. Additionally, I would like to ask whether I can directly use the original PSCATTER_GATHER_LIST m_CopySGL ? Is there a particular reason why you suggested splitting it into m_ListHeader and m_Elements?

Reply:
> @ybendito I would like to discuss some implementation details with you.
> 
> 1. CNdisSharedMemory* m_UsedPages[MAX_PACKET_PAGES]; // or use the list? just for discussion
>    ok, I will change it to a list. At the same time, I will remove the CNdisList<CNdisSharedMemory, CLockedAccess, CCountingObject> SendCopyPagesInUsed;. The original reason I considered using an array was that I was concerned about the possibility of an element being inserted into two lists simultaneously, which could cause errors. Now that SendCopyPagesInUsed has been removed, this issue no longer exists.

If you keep all the pages in the list around specific CNB you do not need SendCopyPagesInUsed at all, do you?
But no problem, both array and list are ok.
There is several things that do require some attention, I'll describe them in a separate comment.

> 
> 2.Regarding the struct CExtendedNBStorage, I understand that this structure is only meant to hold members and will not contain related functions（Except for the destructor and constructor）. For example, MapCopyDataToVirtioSGL will not be included in this struct, correct?

Yes, I just wanted to keep the basic CNB as small as possible and move all the additional fields that we usually do not need to separate allocatable structure.
> 
> 3. SCATTER_GATHER_LIST m_ListHeader // SCATTER_GATHER_LIST
>    SCATTER_GATHER_ELEMENT m_Elements[MAX_PACKET_PAGES]
>    In your example, I see that there are these two members. Since SCATTER_GATHER_LIST includes a flexible array, it seems, at first glance, that the definition of m_ListHeader should be placed at the end; otherwise, it fail to compile. Additionally, I would like to ask whether I can directly use the original PSCATTER_GATHER_LIST m_CopySGL ? Is there a particular reason why you suggested splitting it into m_ListHeader and m_Elements?

actually we do not need the SCATTER_GATHER_LIST, what we need is just SCATTER_GATHER_ELEMENT[MAX_PACKET_PAGES] and the number of the allocated pages, correct?